Tests/BSONTests/ExtendedJSONConversionTests.swift

```diff
@@ -24,5 +59,260 @@ open class ExtendedJSONConversionTestCase: BSONTestCase {
+
+        // Error cases
+        expect(try Int64(fromExtJSON: ["$numberLong": 5], keyPath: []))
+            .to(throwError(errorType: DecodingError.self))
+        expect(try Int64(fromExtJSON: ["$numberLong": "5", "extra": true], keyPath: []))
+            .to(throwError(errorType: DecodingError.self))
+        expect(try Int64(fromExtJSON: ["$numberLong": .string("\(Double(Int64.max) + 1)")], keyPath: ["key", "path"]))
+            .to(throwError(errorType: DecodingError.self))
+    }
+
+    /// Tests the BSON Double [finite] and Double [non-finite] types.
+    func testDouble() throws {
+        // Success cases
+        expect(try Double(fromExtJSON: 5.5, keyPath: [])).to(equal(5.5))
+        expect(try Double(fromExtJSON: ["$numberDouble": "5.5"], keyPath: [])).to(equal(5.5))
+        expect(try Double(fromExtJSON: ["$numberDouble": "Infinity"], keyPath: [])).to(equal(Double.infinity))
+        expect(try Double(fromExtJSON: ["$numberDouble": "-Infinity"], keyPath: [])).to(equal(-Double.infinity))
+        expect(try Double(fromExtJSON: ["$numberDouble": "NaN"], keyPath: [])?.isNaN).to(beTrue())
+
+        // Nil cases
+        expect(try Double(fromExtJSON: JSON.bool(true), keyPath: [])).to(beNil())
+        expect(try Double(fromExtJSON: ["bad": "5.5"], keyPath: [])).to(beNil())
+
+        // Error cases
+        expect(try Double(fromExtJSON: ["$numberDouble": 5.5], keyPath: []))
+            .to(throwError(errorType: DecodingError.self))
+        expect(try Double(fromExtJSON: ["$numberDouble": "5.5", "extra": true], keyPath: []))
+            .to(throwError(errorType: DecodingError.self))
+        expect(try Double(fromExtJSON: ["$numberDouble": JSON.bool(true)], keyPath: ["key", "path"]))
+            .to(throwError(errorType: DecodingError.self))
+    }
+
+    func testDecimal128() throws {
+        // Success cases
+        expect(try BSONDecimal128(fromExtJSON: ["$numberDecimal": "0.020000000000000004"], keyPath: []))
+            .to(equal(try BSONDecimal128("0.020000000000000004")))
+
+        // Nil cases
+        expect(try BSONDecimal128(fromExtJSON: JSON.bool(true), keyPath: [])).to(beNil())
+        expect(try BSONDecimal128(fromExtJSON: ["bad": "5.5"], keyPath: [])).to(beNil())
+
+        // Error cases
+        expect(try BSONDecimal128(fromExtJSON: ["$numberDecimal": 0.020000000000000004], keyPath: []))
+            .to(throwError(errorType: DecodingError.self))
+        expect(try BSONDecimal128(fromExtJSON: ["$numberDecimal": "5.5", "extra": true], keyPath: []))
+            .to(throwError(errorType: DecodingError.self))
+        expect(try BSONDecimal128(fromExtJSON: ["$numberDecimal": JSON.bool(true)], keyPath: ["key", "path"]))
+            .to(throwError(errorType: DecodingError.self))
+    }
+
+    func testBinary() throws {
+        // Success case
+        try expect(try BSONBinary(fromExtJSON: ["$binary": ["base64": "//8=", "subType": "00"]], keyPath: []))
```

Comment:
related to Patrick's previous comment about needing to use `radix: 16`, it might be a good idea to test a larger subtype here with a different value in hex than in decimal.
all of the pre-defined subtypes have values 0-6 so you'd need to use a "user-defined" one, which has to be something in the range of `0x80` to `UInt8.max` I believe